File: src/main/java/com/illuzionzstudios/mist/ui/type/ConfirmUI.java
/**
 * Copyright © 2020 Property of Illuzionz Studios, LLC
 * All rights reserved. No part of this publication may be reproduced, distributed, or
 * transmitted in any form or by any means, including photocopying, recording, or other
 * electronic or mechanical methods, without the prior written permission of the publisher,
 * except in the case of brief quotations embodied in critical reviews and certain other
 * noncommercial uses permitted by copyright law. Any licensing of this software overrides
 * this statement.
 */
package com.illuzionzstudios.mist.ui.type;

import com.cryptomorin.xseries.XMaterial;
import com.illuzionzstudios.mist.config.locale.Locale;
import com.illuzionzstudios.mist.ui.UserInterface;
import com.illuzionzstudios.mist.ui.button.Button;
import com.illuzionzstudios.mist.ui.render.ItemCreator;
import org.bukkit.inventory.ItemStack;

import java.util.function.Consumer;

/**
 * Simple interface to confirm an action
 */
public class ConfirmUI extends UserInterface {

    /**
     * Action to run when after answered.
     * Returns whether was accepted or not
     */
    public final Consumer<Boolean> confirmAction;

    /**
     * Button to deny
     */
    public final Button denyButton;

    /**
     * Button to confirm
     */
    public final Button confirmButton;

    /**
     * @param confirmAction Action to run after confirmed/answered
     */
    public ConfirmUI(Consumer<Boolean> confirmAction) {
        super(null, false);
        setTitle("&8Are you sure?");
        setSize(27);

        this.confirmAction = confirmAction;

        denyButton = Button.of(ItemCreator.builder()
                .material(XMaterial.RED_DYE)
                .name(Locale.Interface.CONFIRM_CONFIRM_NAME)
                .lore(Locale.Interface.CONFIRM_CONFIRM_LORE)
                .glow(true)
                .build(),
                (player, ui, clickType, event) -> {
                    confirmAction.accept(false);
                });

        confirmButton = Button.of(ItemCreator.builder()
                .material(XMaterial.LIME_DYE)
                .name(Locale.Interface.CONFIRM_DENY_NAME)
                .lore(Locale.Interface.CONFIRM_DENY_LORE)
                .glow(true)
                .build(),
                (player, ui, clickType, event) -> {
                    confirmAction.accept(true);
                });
    }

    @Override
    public ItemStack getItemAt(final int slot) {
        if (slot == 11) {
            return confirmButton.getItem();
        } else if (slot == 15) {
            return denyButton.getItem();
        }

        // Else placeholder item
        return ItemCreator.builder().name(" ").material(XMaterial.BLACK_STAINED_GLASS_PANE).build().makeUIItem();
    }

}

File: src/main/kotlin/com/illuzionzstudios/mist/ui/type/ConfirmUI.kt
package com.illuzionzstudios.mist.ui.type

import com.cryptomorin.xseries.XMaterial
import com.illuzionzstudios.mist.config.locale.PluginLocale
import com.illuzionzstudios.mist.item.ItemCreator
import com.illuzionzstudios.mist.ui.UserInterface
import com.illuzionzstudios.mist.ui.button.Button
import com.illuzionzstudios.mist.ui.button.Button.ButtonListener
import org.bukkit.entity.Player
import org.bukkit.event.inventory.ClickType
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.inventory.ItemStack
import java.util.function.Consumer

/**
 * Simple interface to confirm an action
 */
class ConfirmUI(confirmAction: Consumer<Boolean>) : UserInterface(null, false) {
    /**
     * Action to run when after answered.
     * Returns whether was accepted or not
     */
    val confirmAction: Consumer<Boolean>

    /**
     * Button to deny
     */
    val denyButton: Button

    /**
     * Button to confirm
     */
    val confirmButton: Button
    override fun getItemAt(slot: Int): ItemStack? {
        if (slot == 11) {
            return confirmButton.item
        } else if (slot == 15) {
            return denyButton.item
        }

        // Else placeholder item
        return ItemCreator.builder().name(" ").material(XMaterial.BLACK_STAINED_GLASS_PANE).build().makeUIItem()
    }

    /**
     * @param confirmAction Action to run after confirmed/answered
     */
    init {
        title = "&8Are you sure?"
        size = 27
        this.confirmAction = confirmAction
        denyButton = Button.Companion.of(ItemCreator.builder()
            .material(XMaterial.RED_DYE)
            .name(PluginLocale.Companion.INTERFACE_CONFIRM_CONFIRM_NAME.toString())
            .lore(PluginLocale.Companion.INTERFACE_CONFIRM_CONFIRM_LORE.toString())
            .glow(true)
            .build(),
            ButtonListener { player: Player?, ui: UserInterface?, clickType: ClickType?, event: InventoryClickEvent? ->
                confirmAction.accept(
                    false
                )
            })
        confirmButton = Button.Companion.of(ItemCreator.builder()
            .material(XMaterial.LIME_DYE)
            .name(PluginLocale.Companion.INTERFACE_CONFIRM_DENY_NAME.toString())
            .lore(PluginLocale.Companion.INTERFACE_CONFIRM_DENY_LORE.toString())
            .glow(true)
            .build(),
            ButtonListener { player: Player?, ui: UserInterface?, clickType: ClickType?, event: InventoryClickEvent? ->
                confirmAction.accept(
                    true
                )
            })
    }
}